Cross buffer copy

abadams opened the pull request:
Add the ability to explicitly schedule when and how host<->device buffer copies occur. See the new generator test for example usage.

This is done by using a Func to represent the copy. The Func is stored in the destination memory space (only), and the computation it does is the act of copying. The Func must have a definition of the form:

foo(\_) = bar(\_);

The Funcs returned by Func::in() are suitable. You then say foo.copy_to_host() or foo.copy_to_device() in the schedule, which specifies that the Func's "computation", such as it is, is done by a call to a new runtime function which copies from one buffer_t to another (whether on host or device). This is implemented by rewriting its definition to be an extern stage that calls this new cross-device buffer copy runtime function.

Several new mechanisms are added to make this work:

- The PR adds support for extern stages with split store_at/compute_at with GPU outputs. This will be necessary for doing device<->host copies into circular buffers asynchronously. Previously these just failed and leaked in mysterious ways. To do this we need a way to create aliased crops of device allocations, and another call to clean up any resources associated with said crops. Two new calls are added to the device interface to do this. OpenCL and CUDA support for these is added in the PR. Cuda is just pointer math. OpenCL requires a call to clCreateSubBuffer. The ability to crop GPU allocations is not exposed via Halide::Runtime::Buffer yet.

- The PR introduces the idea of proxy expressions for extern stages. Not exposed to the front-end, but we could imagine doing so. An extern stage with a proxy expression uses the proxy for all bounds computation purposes instead of making any dynamic bounds queries. The idea is that whenever we have a Func that gets rewritten by the schedule to use an equivalent extern stage, we'll move its original definition into the proxy expression for bounds purposes.

The ultimate goal of this work is to combine it with async to overlap communication and computation, via asynchronous host<->device copies writing into a circular buffer. For example the following schedule would stage copies of a host producer into a device consumer asyncronously through a device-side circular buffer of 8 scanlines:

    producer.in(consumer).copy_to_device().async().store_root().compute_at(consumer, y).fold_storage(y, 8);

SanderVocke commented:
Q: does this also use bounds inference to copy only the part of a buffer that is actually going to be consumed / has just been produced? Or is the full buffer always copied?

abadams commented:
In this design, no buffer ever exists on both host and dev simultaneously, so yes, the bounds can all be different (because the host and dev buffers are different Funcs!), and are all determined by bounds inference.

abadams commented:
builds are finally green. ptal

dsharletg commented on src/runtime/device_interface.cpp:
```diff
@@ -387,4 +390,176 @@ WEAK void halide_device_and_host_free_as_destructor(void *user_context, void *ob
+    if (dst_device_interface &&
+        dst_device_interface != dst->device_interface) {
+        err = halide_device_free(user_context, dst);
+        err = err || halide_device_malloc(user_context, dst, dst_device_interface);
```
I think this will force the error code to become a bool (losing the error code).

abadams replied:
Fixed.

dsharletg replied:
Is it really fixed? err || x will be true or false, losing the error code stored in err.

dsharletg commented on src/runtime/device_interface.cpp:
```diff
@@ -387,4 +390,176 @@ WEAK void halide_device_and_host_free_as_destructor(void *user_context, void *ob
+        return 0;
+    }
+
+    {
```
The extra scope here is unnecessary

zvookin replied:
I'd probably change the sense of the above if to only run the code if there is a device. It also means the function would end with a return statement, which is good practice.

dsharletg commented on src/runtime/opencl.cpp:
```diff
@@ -1082,6 +1053,38 @@ WEAK uintptr_t halide_opencl_get_cl_mem(void *user_context, halide_buffer_t *buf
+    }
+    offset *= src->type.bytes();
+    cl_buffer_region region = {offset, dst->size_in_bytes()};
+    dst->device = (uint64_t)clCreateSubBuffer((cl_mem)(src->device),
```
Probably worth a comment that the sub-buffer is the linear range of addresses that span the crop.

Does this have any implications for correctness in the other parts of this change?

abadams replied:
Added a comment. That's the expected behavior - the strides don't change when you crop. Really I just want to bump the min pointer to the top-left of the crop region. This is a way to do that in opencl.

abadams commented on src/runtime/device_interface.cpp:
```diff
@@ -387,4 +390,176 @@ WEAK void halide_device_and_host_free_as_destructor(void *user_context, void *ob
+        return 0;
+    }
+
+    {
```
Stylistically, I like putting scoped mutex locks at the top of a scope, so that it's clear to the reader what the mutex encompasses.

abadams commented on src/runtime/HalideRuntime.h:
```diff
@@ -541,6 +546,38 @@ extern int halide_copy_to_host(void *user_context, struct halide_buffer_t *buf);
+ * supported by some device APIs (others will return
+ * halide_error_code_device_crop_unsupported). Do not call device_free
+ * on the resulting buffer. Call halide_device_release_crop instead to
+ * clean up any resources associated with the cropped view. Do not
```
Comment fixed. I want to keep things consistent with wrap/detach device_and_host_malloc/device_and_host_free and device_malloc/device_free for now, which use separate pairs. Buffer currently messes up these cases. It needs an enum to track what it should do with the device pointer on destruction.

abadams commented:
Review ping (or maybe dillon's review counts as an lgtm?)

abadams commented on src/ScheduleFunctions.cpp:
```diff
@@ -496,6 +498,40 @@ Stmt build_produce(Function f, const Target &target) {
+
+            // We have to anticipate failures in the extern stage, so
+            // call this by registering a destructor before the extern
+            // call and triggering it afterwards using an Allocate node.
```
register_destructor can only trigger on function exit, and usually only on failure - I want RAII over a specific scope. Allocate does that.

abadams commented on src/runtime/buffer_t.cpp:
```diff
@@ -153,10 +154,60 @@ halide_buffer_t *_halide_buffer_crop(halide_buffer_t *dst,
+// with distinct store_at/compute_at levels. Each call to the stage
+// only fills in part of the buffer.
+HALIDE_BUFFER_HELPER_ATTRS
+void _halide_buffer_retire_crop(void *user_context,
```
Renamed to _halide_buffer_retire_crop_after_extern_stage

abadams commented:
Looks like I missed one err = err ||. Fixed

dsharletg commented:
Is this ready to be merged?

zvookin commented on Makefile:
```diff
@@ -714,6 +714,7 @@ $(INCLUDE_DIR)/HalideBuffer.h: $(SRC_DIR)/runtime/HalideBuffer.h
 	cp $< $(INCLUDE_DIR)/
 
 $(BIN_DIR)/build_halide_h: $(ROOT_DIR)/tools/build_halide_h.cpp
+	@-mkdir -p $(BIN_DIR)
```
Pattern used currently for mkdirs is "$(@D)" which means "directory part of the target path."

zvookin commented on src/AddImageChecks.cpp:
```diff
@@ -211,7 +211,8 @@ Stmt add_image_checks(Stmt s,
             for (size_t i = 0; i < order.size(); i++) {
                 Function f = env.find(order[i])->second;
-                if (f.has_extern_definition()) {
+                if (f.has_extern_definition() &&
```
This combination is showing up in a number of places in this PR and it may be more readable to introduce a single well named method to return what this is testing for.

zvookin commented on src/CodeGen_LLVM.cpp:
```diff
@@ -2097,7 +2098,7 @@ void CodeGen_LLVM::visit(const Call *op) {
 
         Value *buffer = codegen(op->args[2]);
-        buffer = builder->CreatePointerCast(buffer, buffer_t_type->getPointerTo());
+        buffer = builder->CreatePointerCast(buffer, debug_to_file->getFunctionType()->getParamType(3));
```
I'm slightly dubious about this as it seems to be saying "cast the argument to whatever type the slot has" instead of casting to a type we know the item is. I don't fully understand the context so I may be wrong, but am concerned this may mask an error in the compiler if we are not careful.

abadams replied:
Under certain circumstances I was getting multiple copies of struct halide_buffer_t coming in under different names in the LLVM IR. Diagnosing seemed like a rabbit-hole, so I hit it with a hammer.

zvookin commented on src/Function.h:
```diff
@@ -211,6 +213,15 @@ class Function {
+     * stage. For most Functions, including those with extern
+     * definitions, this will be an undefined Expr. */
+    // @{
+    EXPORT Expr extern_definition_proxy_expr() const;
```
Is it possible to be a bit more explicit re: what "same data access pattern" implies re: requirements on this Expr? E.g. does it mean the bounds of reads from inputs and writes to outputs from the extern stage are no larger than what bounds inference will infer? Clearly it is not possible to make the Expr and the extern have exactly the same set of loads and stores...

abadams replied:
Exactly the same is the current meaning. It could be relaxed to be a statement about the bounds. I'll tweak the comment.

zvookin commented on src/Generator.h:
```diff
@@ -1363,6 +1363,18 @@ class GeneratorInput_Buffer : public GeneratorInputImpl<T, Func>, public Constra
         return *this;
     }
+
+    Func in() {
```
Shouldn't this just go in a trivial PR or submit to master? @steven-johnson

abadams replied:
Wasn't sure what we've decided w.r.t forwarding methods from ```Input<Func>``` I'll wait to see what Steven says.

zvookin commented on src/Generator.h:
```diff
@@ -1438,6 +1450,18 @@ class GeneratorInput_Func : public GeneratorInputImpl<T, Func> {
         this->estimate_impl(var, min, extent);
         return *this;
     }
+
```
Ditto above.

zvookin commented on src/runtime/HalideRuntime.h:
```diff
@@ -541,6 +546,38 @@ extern int halide_copy_to_host(void *user_context, struct halide_buffer_t *buf);
                                  const struct halide_device_interface_t *device_interface);
 
+/** Copy data from one buffer to another. The buffers may have
+ * different shapes and sizes, but the destination buffer's shape must
```
Might be good to be clear here that only the mins and extents count. I assume the strides can be arbitrarily different.

zvookin commented on src/runtime/HalideRuntime.h:
```diff
@@ -541,6 +546,38 @@ extern int halide_copy_to_host(void *user_context, struct halide_buffer_t *buf);
 
+/** Copy data from one buffer to another. The buffers may have
+ * different shapes and sizes, but the destination buffer's shape must
+ * be contained within the source buffer's shape. The source data is
```
We should explicitly spell out the preconditions and postconditions on src and dst state. The completely general API allows the initial state to be almost anything. The src can be on any or no device and if it does have a device, may have a null host pointer. Basically the only requirement would be that at least one thing is non-null. (Likely with the caveat that a zero extent copy is a nop no matter what.) The dirty bits can be any of the three valid states, and the dst can be on any device or no device and can probably even have a null host pointer, in which case we allocate one if dst_device_interface is null. The latter requires the caller to call halide_free on the host pointer if it is allocated. The final state of the dirty bits would also be somewhat wide open, though some specification can be made.

In the above, the only thing that I don't see being easily doable is a combined device and host allocation as there is no way to control whether one wants one or not. We could assume one always does if it is available and the dst buffer does not already have some sort of allocation. Note that specifying all of this is important because the caller will have to make sepcific deallocation calls to cleanup the result.

All of that is both a bit of work to specify and will absolutely require test coverage as the internals are non-trivial.

Perhaps the most restricted interface insists that the source and destination allocations already exist and the other side of each buffer is not dirty. In this case, one would probably just have a bool to indicate whether copying to the host or device side of the destination.

I am aware that this borders on being pedantic, but I really do feel that if we aren't rigorous about specifying this call, it will cause problems.

zvookin commented on src/runtime/HalideRuntime.h:
```diff
@@ -541,6 +546,38 @@ extern int halide_copy_to_host(void *user_context, struct halide_buffer_t *buf);
+ * halide_error_code_device_crop_unsupported). Call
+ * halide_device_release_crop instead of halide_device_free to clean
+ * up resources associated with the cropped view. Do not call
+ * device_free on the source buffer while the destination buffer still
```
Probably should say the parent cannot be freed by any means. (It could be a combined allocation or a wrapped native, etc.)

zvookin commented on src/runtime/HalideRuntime.h:
```diff
@@ -541,6 +546,38 @@ extern int halide_copy_to_host(void *user_context, struct halide_buffer_t *buf);
+
+/** Give the destination buffer a device allocation which is an alias
+ * for the same coordinate range in the source buffer. Modifies the
+ * device, device_interface, and the device_dirty flag only. Only
```
Why would cropping modify the device dirty flag?

abadams replied:
If you create a crop of a dirty buffer, the crop is also dirty.

zvookin commented on src/runtime/HalideRuntime.h:
```diff
@@ -541,6 +546,38 @@ extern int halide_copy_to_host(void *user_context, struct halide_buffer_t *buf);
+ * pulled from either device or host memory on the source, depending
+ * on the dirty flags. host is preferred if both are valid. The
+ * dst_device_interface parameter controls the destination memory
+ * space. NULL means host memory. */
```
We should also probably state explicitly whether halide_buffer_copy and halide_buffer_crop are allowed to modify the host memory in device to device cases and such.

It might also be useful to model halide_copy_to_host and halide_copy_to_device in terms of halide_buffer_copy and provide that mapping for illustrative purposes.

zvookin commented on src/runtime/HalideRuntime.h:
```diff
@@ -541,6 +546,38 @@ extern int halide_copy_to_host(void *user_context, struct halide_buffer_t *buf);
+ * on the dirty flags. host is preferred if both are valid. The
+ * dst_device_interface parameter controls the destination memory
+ * space. NULL means host memory. */
+extern int halide_buffer_copy(void *user_context, struct halide_buffer_t *src,
```
Is it a nop or an error for src and dst to be the same buffer?

zvookin commented on src/runtime/HalideRuntime.h:
```diff
@@ -541,6 +546,38 @@ extern int halide_copy_to_host(void *user_context, struct halide_buffer_t *buf);
+ * up resources associated with the cropped view. Do not call
+ * device_free on the source buffer while the destination buffer still
+ * lives.  Note that the two buffers do not share dirty flags, so care
+ * must be taken to update them together as needed. Note also that
```
I'm a bit dubious about the last part. We will likely solve this in Halide::Runtime::Buffer by always translating the crop back into the parent's coordinates and cropping from there. (This is tricky in the general case as the mins could be reset on one of the buffers. One has to be able to determine the relative offset of the device memory references I think...) In any event, we should explicitly state that multiple crops can be taken from the same parent buffer.

If crop from a previous crop is not allowed on some architectures, I'm tempted to force all devices to make it an error for portability reasons. (Though realistically, I think we should investigate getting rid of the restriction.)

zvookin commented on src/runtime/HalideRuntime.h:
```diff
@@ -541,6 +546,38 @@ extern int halide_copy_to_host(void *user_context, struct halide_buffer_t *buf);
+/** Copy data from one buffer to another. The buffers may have
+ * different shapes and sizes, but the destination buffer's shape must
+ * be contained within the source buffer's shape. The source data is
+ * pulled from either device or host memory on the source, depending
```
Another complication: if the call can deallocate a device handle, that's complicated as the handle could be a halide_device_malloc result, a combined device/host allocation, a wrapped_native, or a crop. This is already a bug in halide_copy_to_device when it transitions the buffer to a new device. I expect one can trigger this by passing an output buffer that is e.g. an OpenGL wrapped native, and having the output Func scheduled cuda....

abadams replied:
Hrm, everywhere we flip buffers from one device to another has the same problem - we just call device_free. Perhaps these should be errors instead?

zvookin commented on src/runtime/buffer_t.cpp:
```diff
@@ -153,10 +154,60 @@ halide_buffer_t *_halide_buffer_crop(halide_buffer_t *dst,
     }
-    dst->host += offset * src->type.bytes();
+    offset *= src->type.bytes();
+    dst->host += offset;
```
We need to conditionalize this on non-null host allocation. (Can show up for device only buffers, but possibly also for lazy allocated ones.) If not, it should probably assert non-null. Adding to null pointers makes for difficult to track down bugs. (Also: https://github.com/halide/Halide/issues/2058 .)

abadams commented:
For now I removed all device flipping behavior from device_interface.cpp, and replaced it with equivalent logic (for internal allocations!) in InjectHostDevBufferCopies.

It's not just that we don't know which free method to use (or what to do if it's zero-copy), it's also that the pipeline doesn't understand any reference counting that might be going on at the Halide::Runtime::Buffer level. 

With the latest few commits, if you try to pass an input image with an opencl device backing to a pipeline that wants to give it a cuda device allocation, you now get an error.

zvookin commented:
One thing we could do is have bounds query fill in the device interface to indicate on which device the allocation has to be...
